### ISSUE TITLE
no-did-mount-set-state: not triggered when `this` is not explicit

### ISSUE THREAD

**Issue by caesarsol** [yannickcr/eslint-plugin-react#804]
Example of code that does not trigger rule:

```
class Compo extends React.Component {
  componentDidMount() {
    const self = this;
    self.setState({a: 1});
  }
}
```

I don't know if it's possible to follow the references to `this`?

**Comment by ljharb** [comment]
This seems reasonable - but arrow functions exist - why do you need the `var self = this` pattern?

**Comment by caesarsol** [comment]
There are cases in which you need both inner `this` and outer `this`, for example in some event listeners.

Actually I never use that pattern, but happened that someone else used it on the code I'm working on and that prevented the linter to highlight the error.

Also, this project does not explicitly state that we are using ES6, in some environments arrow functions could be off limits.

(I understand it's not a trivial change to support this, I'm just checking! :) )

**Comment by ljharb** [comment]
Which cases? Event handlers should be referencing elements using `event.target` and friends, not with `this`.

If you're using `class`, then yes, you are explicitly using ES6 :-p

**Comment by ljharb** [comment]
Since there's literally no justification for using the `self = this` pattern when arrow functions are available, and `class` usage guarantees that arrow functions are available, I don't think there's any value to add complexity here.
